libsnark/gadgetlib1/gadgets/hashes/anemoi/tests/test_anemoi_gadget.cpp

```diff
@@ -339,31 +350,115 @@ void test_anemoi_permutation_mds()
 {
-    // Execute all tests for the given curve.
-
-    ppT::init_public_params();
     // Use debug parameters with small values for the small gadgets
     using parameters_debug = parameters_debug_bls12_381;
     test_flystel_Q_gamma_prime_field_gadget<ppT, parameters_debug>();
     test_flystel_Q_gamma_binary_field_gadget<ppT, parameters_debug>();
     test_flystel_E_power_five_gadget<ppT>();
     test_flystel_E_root_five_gadget<ppT, parameters_debug>();
     test_flystel_prime_field_gadget<ppT, parameters_debug>();
+    test_anemoi_permutation_mds_bls12_381();
+}
+
+template<typename ppT>
+void test_for_curve(
+    expected_round_values_fn_t<ppT> expected_round_values_fn = 0)
+{
     // Use the original parameters for the full permutation
     using parameters = anemoi_parameters<ppT>;
-    test_anemoi_permutation_round_prime_field_gadget<ppT, 1, parameters>();
-    test_anemoi_permutation_round_prime_field_gadget<ppT, 2, parameters>();
-    test_anemoi_permutation_round_prime_field_gadget<ppT, 3, parameters>();
-    test_anemoi_permutation_round_prime_field_gadget<ppT, 4, parameters>();
-    test_anemoi_permutation_mds<ppT, parameters>();
+    test_anemoi_permutation_round_prime_field_gadget<ppT, 1, parameters>(
+        expected_round_values_fn);
+    test_anemoi_permutation_round_prime_field_gadget<ppT, 2, parameters>(
+        expected_round_values_fn);
+    test_anemoi_permutation_round_prime_field_gadget<ppT, 3, parameters>(
+        expected_round_values_fn);
+    test_anemoi_permutation_round_prime_field_gadget<ppT, 4, parameters>(
+        expected_round_values_fn);
+}
+
+template<typename ppT> void test_curve_parameters()
+{
+    using parameters = anemoi_parameters<ppT>;
+    printf("g = %zd\n", parameters::multiplicative_generator_g);
+    printf("alpha = %zd\n", parameters::alpha);
+    printf("beta = %zd\n", parameters::beta);
+    printf("gamma = %zd\n", parameters::gamma);
+    printf("quad_exponent = %zd\n", parameters::quad_exponent);
+    printf("alpha_inv = ");
+    parameters::alpha_inv.print();
+    printf("delta = ");
+    parameters::delta.print();
+}
+
+TEST(TestAnemoiGadget, BLS12_381)
+{
+    test_intermediate_gadgets_bls12_381<libff::bls12_381_pp>();
 }
 
-TEST(TestAnemoiGadget, BLS12_381) { test_for_curve<libff::bls12_381_pp>(); }
+TEST(TestCurveParameters, BLS12_381)
+{
+    test_curve_parameters<libff::bls12_381_pp>();
+    test_for_curve<libff::bls12_381_pp>(&anemoi_expected_output_one_round);
+}
+
+TEST(TestCurveParameters, BLS12_377)
+{
+    test_curve_parameters<libff::bls12_377_pp>();
+    // TODO For BLS12_377 alpha = 11, which is the first value for
+    // which alpha is co-prime to r-1, required for the inverse
+    // alpha^-1 to exist (r is the modulus of Fr). ATM we have a gadget
+    // only for alpha = 5 (flystel_E_power_five_gadget), but not for
+    // alpha = 11. For this reason test_for_curve does not run on
+    // BLS12_377.
+    // test_for_curve<libff::bls12_377_pp>();
```

Comment:
Can this not be called?

Reply:
As the comment explains: "For `BLS12_377` (and `MNT6`) we have `alpha = 11`, which is the first value for which `alpha` is co-prime to `r-1`, required for the inverse `alpha^-1` to exist (`r` is the modulus of `Fr`). ATM we have a gadget only for `alpha = 5` (`flystel_E_power_five_gadget`), but not for `alpha = 11`. For this reason test_for_curve does not run on BLS12_377." . Should we implement a gadget for `alpha = 11` at this stage, do you think?

Reply:
OK.  So 5 works for all the other curves except these.  Understood.